### DISCUSSION TITLE
String ID

### DISCUSSION THREAD

**Discussion by gustavorozolin** [isar/isar#344] [Help]
My app have String ID for all entities (Server side) I want migrate from [Hive](https://pub.dev/packages/hive) to [Isar](https://pub.dev/packages/isar) so I was reading the [Isar documentation](https://isar.dev/tutorials/quickstart.html) and find out it doesn't support String ID (only `int`), Yes I know that I can create a Class like this:

```
@Collection()
class Student {
  int? id;
  @Index(unique: true)
  String? myServerId;
  late String name;
  final teacher = IsarLink<Teacher>();
}
```
but when I fetch data (`JSON`) from server the `id` field (manager for **Isar**) don't came from server, so I save this record local **Isar** will create the ID OK, next time I get the same record from server is needed check via `myServerId` and if exist just update with id created by **Isar** its easy, how I can manager its with all entities fields like `teacher`? Need check every entities child or have a better way? Thanks in advance

**Comment by renatoromanini** [comment]
Up

**Comment by simc** [comment]
Hey @renatoromanini and @gustavorozolin,

It was an intentional decision to use integers as internal id because it allows some significant performance improvements for storing links and indexes. If you want to use a string id, it is very easy to do: add a unique index to your string id field and use the generated `collection.getByYourId()`, `collection.deleteByYourId()` etc. methods 👍 

Using an index offers the same performance as using the integer id.

**Reply by gustavorozolin** [reply]
Thanks for your quick reply @leisim, but I will need make this check for each children of my entity? For example if I update a student object I have to check student ID and teacher ID...?

**Reply by simc** [reply]
I'm not sure what you mean exactly. Could you please provide a small example?

**Comment by gustavorozolin** [comment]
imagine the following scenario:

```
@Collection()
class Teacher {
  int? id;
  @Index(unique: true)
  String? myServerId;
  late String subject;
}

@Collection()
class Student {
  int? id;
  @Index(unique: true)
  String? myServerId;
  late String name;
  final teacher = IsarLink<Teacher>();
}
```
So I fetch my data from server: 

`{myServerId: 'id123', name: 'Jhon', teacher: {mySeerverId: 'id1234', subject: 'Math'}}`

Convert `JSON` to objects and save, the first time that I saved this object **Isar** will create a `id` for `Student` and `Teacher` thats ok, but when I fetch this same `Json` from server again when I will save,  I need check `myServerId` from `Student` and `Teacher` because I check only `Student` I will get a exception cause `Teacher` with 'id1234' already exists. Yes I know that I can write a code that check `myServerId` for each `Isar's @Collection()` but I believe its not good for performance.

**Reply by ghost** [reply]
If I understand correctly, you can use ```put(object, replaceOnConflict: true)```. 

First check if one exists in Isar with `getById`, and then `put` a new object from the response:

```put(Object.fromJson({... response.toJson, "isarId": _localObject?.isarId}), replaceOnConflict: true)```

**Reply by simc** [reply]
Yes exactly!

**Reply by gustavorozolin** [reply]
I ran this code:

```
String json = '{"myServerId": "id123", "name": "Jhon", "teacher": {"mySeerverId": "id1234", "subject": "Math"}}';

    await isar!.writeTxn((isar) async {
      for(int i = 0; i < 10000; i++) {
        Student student = studentFromJson(json);
        student.name = student.name! +' '+i.toString();
        await isar.students.put(student, replaceOnConflict: true, saveLinks: true);
      }
    });


    var count = await isar!.teachers.countSync();
    print(count);

```
and I got an error: `[VERBOSE-2:ui_dart_state.cc(209)] Unhandled Exception: IsarError: Unique index violated.`

But when I run without `saveLinks: true` I get no errors and don't save (of course) teacher, its look like `replaceOnConflict: true` affect only the main `@Collection()`

**Reply by simc** [reply]
Thanks for the report: https://github.com/isar/isar/issues/354

**Reply by gustavorozolin** [reply]
@leisim and @lodisy  thanks for the quick support

**Comment by JCKodel** [comment]
> It was an intentional decision to use integers as internal id because it allows some significant performance improvements for storing links and indexes.

> Using an index *offers the same performance as using the integer id*.

O.o

That should be a choice. For offline-first applications, Isar is unusable (since we can't rely on int keys that will clash some eventual sync with a server).

**Reply by frank06** [reply]
Not really. Have you read this? https://isar.dev/recipes/string_ids.html

**Reply by JCKodel** [reply]
> Not really. Have you read this? https://isar.dev/recipes/string_ids.html

You cannot be serious about hashing a String (considering that what I'm saying is PRECISELY to avoid collision).

Anyway, the lack of web support and the "only-local-single-user database" (because of this issue) makes Isar unusable for most scenarios =\

**Reply by tomekit** [reply]
I've got even faster hashing function: `Id get isarId => stringId.length;` for String IDs !

... but seriously relying on 64-bit int is pretty dangerous.

 For most apps it may not matter, but once the number of rows grows so are the chances. Here is the nice read: https://lemire.me/blog/2019/12/12/are-64-bit-random-identifiers-free-from-collision/

**Reply by JCKodel** [reply]
Good news: Isar 4.0 dev 14 already accepts strings as keys (and they can even be read-only, such as this example I've done for a login entry:)

```dart
  String get id => "${lastLogin.year.toRadixString(16)}"
      "${lastLogin.month.toRadixString(16)}"
      "${lastLogin.day.toRadixString(16).padLeft(2, '0')}"
      ":${userId}";
```

Basically, this table only inserts new data for an user once a day, so this id is perfect.

For any other table, Isar 4.0 now uses String as id.

It also supports SQLite as backend (with encryption), making it more compatible with web (I guess).

**Reply by tomekit** [reply]
Thanks for letting me know. That's great that Isar improved in these areas, looking forward to stable release.

Speaking of your example, do you need ID to efficiently update your rows later or do you need some row grouping (hence the somewhat exotic year/month/day extract logic)?
Wouldn't it be more straightforward to create compound unique index on `userId` and `lastLogin` (Datetime)? - https://isar.dev/indexes.html#composite-indexes

**Reply by JCKodel** [reply]
It's a login log. My intent is to register when the user opened the app, but with 1 day granularity. The row is never updated. It could be simpler (maybe a shorter epoch, such as milliseconds since Jan 1 2023).

If the user logs in more than once a day, the last data is overwritten and that's fine.

Since there will be no row scan (other than server sync stuff), no need for indexes.

I could generate a UUID (I'm using `ksuid` package for that), but then I would need to search the today's log to get the id and replace it...
